homeassistant/components/mqtt.py

```diff
@@ -0,0 +1,164 @@
+                                     'client_id']},
+                           _LOGGER):
+        return False
+
+    try:
+        import paho.mqtt.client as mqtt
+
+    except ImportError:
+        _LOGGER.error("Error while importing dependency paho-mqtt.")
+
+        return False
+
+    global MQTT_CLIENT
+    MQTT_CLIENT = MQTT(config[DOMAIN]['broker'],
+                       config[DOMAIN]['port'],
+                       config[DOMAIN]['topic'],
+                       config[DOMAIN]['keepalive'],
+                       config[DOMAIN]['client_id'])
+
+    def stop_mqtt(event):
+        """ Stop MQTT component. """
+        MQTT_CLIENT.stop()
+
+    def start_mqtt(event):
+        """ Launch MQTT component when Home Assistant starts up. """
+        try:
+            MQTT_CLIENT.run()
+
+        except ConnectionRefusedError:
+            _LOGGER.error("Can't connect to the broker. "
+                          "Please check your settings and the broker itself.")
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_mqtt)
+
+    def send_message(call):
+        """ Sending an MQTT message. """
+        subtopic = 'master'
+        complete_topic = 'home-assistant/{}'.format(str(subtopic))
+        MQTT_CLIENT.publish(complete_topic, str(call))
+
+    def received_mqtt(call):
+        """ Action when an MQTT message arrives. """
+        pass
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_mqtt)
+
+    hass.services.register(DOMAIN, MQTT_SEND, send_message)
+    hass.services.register(DOMAIN, MQTT_RECEIVED, received_mqtt)
+
+    return True
+
+
+# This is based on one of the paho-mqtt examples:
+# http://git.eclipse.org/c/paho/org.eclipse.paho.mqtt.python.git/tree/examples/sub-class.py
+class MQTT(object):
+    """ Implements messaging service for MQTT. """
+    def __init__(self, broker, port, topic, keepalive, clientid=None):
+
+        import paho.mqtt.client as mqtt
+
+        self._broker = broker
+        self._port = port
+        self._topic = topic
+        self._keepalive = keepalive
+        self.msg = None
+
+        self._mqttc = mqtt.Client(clientid)
+        self._mqttc.on_message = self.mqtt_on_message
+        self._mqttc.on_connect = self.mqtt_on_connect
+        self._mqttc.on_publish = self.mqtt_on_publish
+        self._mqttc.on_subscribe = self.mqtt_on_subscribe
+
+    def mqtt_on_connect(self, mqttc, obj, flags):
+        """ Connect callback """
+        _LOGGER.info('Connected to broker %s', self._broker)
+
+    def mqtt_on_publish(self, mqttc, obj, mid):
+        """ Publish callback """
+        pass
+
+    def mqtt_on_subscribe(self, mqttc, obj, mid, granted_qos):
+        """ Subscribe callback"""
+        complete_topic = self._topic + '/#'
+        _LOGGER.info('Subscribed to %s', complete_topic)
+
+    def mqtt_on_message(self, mqttc, obj, msg):
+        """ Message callback """
+        self.msg = '{} {} {}'.format(msg.topic, str(msg.qos), str(msg.payload))
+        print(self.msg)
```

Comment:
I would suggest that any message that we received is send as an event over the bus. This will make it very easy for other components to subscribe.

_(please note that I am not familiar with the MQTT jargon at all and just made the attributes up)_

```python
EVENT_MQTT_MESSAGE_RECEIVED = 'MQTT_MESSAGE_RECEIVED'
hass.event.fire(EVENT_MQTT_MESSAGE_RECEIVED, {
  topic: […],
  subtopic: […],
  message: […],
})
```